File: backend/src/main/java/com/lsoft/aguapotable/service/ComunidadService.java
package com.lsoft.aguapotable.service;

import com.lsoft.aguapotable.domain.entity.Comunidad;
import com.lsoft.aguapotable.domain.repository.ComunidadRepository;
import com.lsoft.aguapotable.dto.request.ComunidadRequest;
import com.lsoft.aguapotable.exception.BusinessException;
import com.lsoft.aguapotable.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ComunidadService {

    private final ComunidadRepository comunidadRepository;

    public List<Comunidad> listarActivas() {
        return comunidadRepository.findByActivoTrue();
    }

    public Comunidad obtenerPorId(Long id) {
        return comunidadRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Comunidad", id));
    }

    @Transactional
    public Comunidad crear(ComunidadRequest request) {
        if (comunidadRepository.existsByNombre(request.getNombre())) {
            throw new BusinessException("Ya existe una comunidad con el nombre: " + request.getNombre());
        }
        Comunidad comunidad = Comunidad.builder()
                .nombre(request.getNombre())
                .direccion(request.getDireccion())
                .telefono(request.getTelefono())
                .email(request.getEmail())
                .ruc(request.getRuc())
                .provincia(request.getProvincia())
                .municipio(request.getMunicipio())
                .activo(true)
                .build();
        return comunidadRepository.save(comunidad);
    }

    @Transactional
    public Comunidad actualizar(Long id, ComunidadRequest request) {
        Comunidad comunidad = obtenerPorId(id);
        comunidad.setNombre(request.getNombre());
        comunidad.setDireccion(request.getDireccion());
        comunidad.setTelefono(request.getTelefono());
        comunidad.setEmail(request.getEmail());
        comunidad.setRuc(request.getRuc());
        comunidad.setProvincia(request.getProvincia());
        comunidad.setMunicipio(request.getMunicipio());
        return comunidadRepository.save(comunidad);
    }

    @Transactional
    public void desactivar(Long id) {
        Comunidad comunidad = obtenerPorId(id);
        comunidad.setActivo(false);
        comunidadRepository.save(comunidad);
    }
}

File: backend/src/comunidad/comunidad.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Comunidad } from '../common/entities';
import { ComunidadRequest } from './dto/comunidad.dto';
import {
  BusinessException,
  ResourceNotFoundException,
} from '../common/exceptions';

@Injectable()
export class ComunidadService {
  constructor(
    @InjectRepository(Comunidad)
    private readonly repo: Repository<Comunidad>,
  ) {}

  async listarActivas(): Promise<Comunidad[]> {
    return this.repo.find({ where: { activo: true } });
  }

  async obtenerPorId(id: number): Promise<Comunidad> {
    const comunidad = await this.repo.findOne({ where: { id } });
    if (!comunidad) throw new ResourceNotFoundException('Comunidad', id);
    return comunidad;
  }

  async crear(dto: ComunidadRequest): Promise<Comunidad> {
    const existe = await this.repo.findOne({
      where: { nombre: dto.nombre },
    });
    if (existe) {
      throw new BusinessException(
        'Ya existe una comunidad con el nombre: ' + dto.nombre,
      );
    }
    const comunidad = this.repo.create({
      nombre: dto.nombre,
      direccion: dto.direccion ?? null,
      telefono: dto.telefono ?? null,
      email: dto.email ?? null,
      ruc: dto.ruc ?? null,
      provincia: dto.provincia ?? null,
      municipio: dto.municipio ?? null,
      activo: true,
    });
    return this.repo.save(comunidad);
  }

  async actualizar(id: number, dto: ComunidadRequest): Promise<Comunidad> {
    const comunidad = await this.obtenerPorId(id);
    comunidad.nombre = dto.nombre;
    comunidad.direccion = dto.direccion ?? null;
    comunidad.telefono = dto.telefono ?? null;
    comunidad.email = dto.email ?? null;
    comunidad.ruc = dto.ruc ?? null;
    comunidad.provincia = dto.provincia ?? null;
    comunidad.municipio = dto.municipio ?? null;
    return this.repo.save(comunidad);
  }

  async desactivar(id: number): Promise<void> {
    const comunidad = await this.obtenerPorId(id);
    comunidad.activo = false;
    await this.repo.save(comunidad);
  }
}
